llvm/lib/Transforms/Scalar/NewGVN.cpp

```diff
@@ -2597,34 +2597,61 @@ bool NewGVN::OpIsSafeForPHIOfOps(Value *V, const BasicBlock *PHIBlock,
   Worklist.push_back(V);
   while (!Worklist.empty()) {
     auto *I = Worklist.pop_back_val();
-    if (!isa<Instruction>(I))
+    if (!(isa<Instruction>(I) || isa<MemoryAccess>(I)))
```

Comment:
I am not clear if you need this generalization or if you can keep the invariant that the worklist only has instructions and MemoryPhis. When checking operands, switch to adding the instruction corresponding to an access that's not a MemoryPhi. For all instructions, check operands for the corresponding MA and for the instruction.

Reply:
I tried something similar but ended up going with this version because I thought the code was easier to understand. I'll update the patch so you can decide which version you prefer.

Reply:
I have updated the PR so that the only MemoryAccesses in the worklist are MemoryPhis.